### ISSUE TITLE
BPE sub-tokenization

### ISSUE THREAD

**Issue by mzeidhassan** [ModernMT/MMT#385]
Hi @davidecaroselli ,

I hope it's all well with you. I see that you are working on 2.5 release :-). I look forward to using this new release and I hope it will have some good improvements. I will be waiting for the official announcement :-) 

I was reading through this issue [https://github.com/ModernMT/MMT/issues/286](url) and I am wondering if there is an easy way to protect terms/product names, etc from BPE tokenization. 

I came across this and they are using a text file to protect phrases using regex. 
https://github.com/daniel-kukiela/nmt-chatbot/blob/master/setup/protected_phrases_bpe.txt

ModernMT is giving really good translation quality, but if there is a way to avoid these made-up words, it would make it the best. 

So, is there a way to utilize a text file to protect particular words/terms/product names, etc based on regex patterns? Would a RegexTokenizer or PatternTokenizer be better in such scenario? What do you think? 

Thanks,
Mohamed

**Comment by davidecaroselli** [comment]
Hi @mzeidhassan 

so, I think that with the 3.0 release (the one with the Transformer), we can think of something like this.

The problem is that with Neural MT you need a fixed size vocabulary, so a regex is not acceptable because you cannot tell the exact words you are protecting. But we can make something different:

1. We reserve 10 tokens that the BPE won't split (i.e. __PROTECTED_1__, __PROTECTED_2__, ...)
2. We create a pre-processor that, with regex, identifies protected tokens and replace them with the keywords __PROTECTED_X__.
3. We create a post-processor that replaces the placeholders __PROTECTED_X__ with the original word.

This way we would be able to protect as many words as we want. This way we should also solve issues like #362 .

**Comment by mzeidhassan** [comment]
Thank you so much @davidecaroselli  for the prompt reply. I appreciate it.

This will be a great enhancement/addition. Having this along with Transformer would change the game. Thanks again @davidecaroselli for your amazing support, as always.

**Comment by davidecaroselli** [comment]
Hi @mzeidhassan ,

no, and for a good reason: Both first and last layers of a Neural Network (embedding and softmax) is proportional in size to the  size of the vocabulary. It is absolutely common to have a vocabulary of ~1.000.000 words for a simple language, and much more for a complex one like German.

If you don't use BPE, you'll end up with a gigantic network that probably won't even fit the GPU RAM. BPE is there for 2 very good reasons:

1. Reduce the vocabulary size from an arbitrary high number of words to a much smaller one still capable of representing the whole language vocabulary (a BPE vocabulary size is usually between 16.000 and 32.000 symbols)
2. Make the network be able to learn new words. With a fixed vocabulary there is absolutely no way the network will be able to learn new word, because it has no way to represent it.

Hope this helps,
Davide

**Comment by mzeidhassan** [comment]
Thanks @davidecaroselli  for your reply.  I see.

These made-up words are very problematic. For example:

En: For example, OSBPARENTOBJ1_C.
JA: たとえば、 OBPoraNTj 1_C.

I think BPE is causing this and I am wondering what can be done to prevent this from happening on the neural network level.

**Comment by mzeidhassan** [comment]
Hi @davidecaroselli 
I believe you are using subword-nmt in the background to create the BPE model. If yes, you may check this [issue](https://github.com/rsennrich/subword-nmt/issues/49). Apparently, there is a glossaries field that can be used to protect specific words from BPE tokenization. I thought I should let you know.

**Comment by mzeidhassan** [comment]
Thanks @LoekvanKooten! Yes, RegEx is always handy, but I am wondering if there is a way to fix this in the BPE sub-tokenization. I am dealing with too many occurrences and RegEx tends to be very slow sometimes. So, a fix from within MMT would be the best.

**Comment by davidecaroselli** [comment]
> Where and how do we define these tag

These tags are already pre-defined, so you can use them directly in your translation request like `This is an example of ${DNT0} ` sentence.

> can this be done with an existing engine

No unfortunately, this needs a re-train of the engine with the new version in `master` branch. Again, I would like to stress the fact that until the final v3.0 release, you should consider this code "in development" and not ready for production use :) .

**Comment by LoekvanKooten** [comment]
Loud and clear! Thank you very much.

**Comment by robertBrnnn** [comment]
Hi @davidecaroselli ,
Great to see the support for DNT has been added, and looking forward to testing it out.

Regarding the number of DNT tokens, it could be quite limiting that there's only ten, depending on the strings that get sent for translation. Would it be possible to increase the amount of DNT tokens? 
i.e. in `/src/decoder-neural/src/main/python/nmmt/t2t/problem.py make_reserved_tokens(self)` change `[('${DNT%d}_' % i) for i in range(10)]` to a higher number? Even if this could be user specified at training time it would allow greater flexibility, or is there a further reason as to why the limit is 10 tokens?

Thanks again,
Rob

**Comment by davidecaroselli** [comment]
> Would it be possible to increase the amount of DNT tokens?
Yes, as you found out you can increase the number of tokens by simply changing `range(10)`. 

However I stil think that 10 can be enough for a simple reason: it is not mandatory for the tokens to be unique in the same sentence. So you can identify the 10th token as the second instance of ${DNT0}, the 11th as the second instance of ${DNT1}, and so on.

If you think this approach doesn't suite your needs, you can proceed by increasing the number of DNT tokens.

Davide

**Comment by LoekvanKooten** [comment]
@davidecaroselli A really noobish question, but if I go to Tensorboard during training and click on graph, all I see is numbers. There should be a setting somewhere that shows you the actual words that are being trained on. How do I get those? Tensorboard is pointing to the tmp/neural_model directory.

**Comment by davidecaroselli** [comment]
Hi all,

this feature has been integrated into v3.0. I'm closing this issue but, if you have any other questions, please don't hesitate to re-open it!
